build: cmake: improve source generated for check_header

tchaikov opened the pull request:
in check_headers.cmake, we verify the self containness of a header file by replicating it and remove `#pragma once` directive in this header. but this approach failed to compile headers which includes a header file with the same name in the root source directory, as we add `-I<directory-of-original-header>` in the cflags when building the generated source file, so that it can include the headers in the same directory. but this confuses the compiler, as, assuming we have "log.hh" in current directory, and under the root source directory, the compiler would always include the "log.hh" in the current directory even it should have included "log.hh" under the root source directory.

in this change, instead of adding `-I<directory-of-original-header>` to cflags, we just include the header under test in a new .cc file created for testing. this should address this problem.

---

this is a cmake-related change, hence no need to backport.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12560/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12560/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4320/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1255/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12343/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12560/)
#### Build Details:
- Duration: 3 hr 50 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12564/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12564/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1257/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4322/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12345/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12564/)
#### Build Details:
- Duration: 3 hr 40 min
- Builder: spider5.cloudius-systems.com

tchaikov commented:
@denesb hi Botond, could you please help review this change as well?

regevran commented:
> in this change, instead of adding -I<directory-of-original-header> to cflags, we just include the header under test in a new .cc file created for testing

The changes in the CMakeFile.txt are not aligned with this description.
Specifically to
> instead of adding -I<directory-of-original-header> to cflags


Also, can you elaborate on how `check_headers` works?
Preferably in the file itself. Preferably with a small concrete example.

tchaikov commented:
> > in this change, instead of adding -I to cflags, we just include the header under test in a new .cc file created for testing
> 
> The changes in the CMakeFile.txt are not aligned with this description. Specifically to
> 
> > instead of adding -I to cflags

it does. please read `list(APPEND includes "${CMAKE_SOURCE_DIR}/${file_dir}")` and understand `includes` is used.

> 
> Also, can you elaborate on how `check_headers` works? Preferably in the file itself. Preferably with a small concrete example.

i'd prefer addressing the comments on this change introduced in this very pull request, and to improve the readability of the code instead of repeating it with a comment.

regevran commented:
>  and to improve the readability of the code

This is achieved by adding comments and examples.
Letting readers getting into the details of the code is not a very good way to communicate your _intents_.
Reading the code communicates what it _does_ - but not what you **think** it suppose to do.
By adding a preface with your intent, and a general approach to this check, the reader can _verify_ that the code you wrote complies to your plans.

tchaikov commented:
> > and to improve the readability of the code
> 
> This is achieved by adding comments and examples. Letting readers getting into the details of the code is not a very good way to communicate your _intents_. Reading the code communicates what it _does_ - but not what you **think** it suppose to do. By adding a preface with your intent, and a general approach to this check, the reader can _verify_ that the code you wrote complies to your plans.

thank you for the comment.
